--- a/mt76_connac_mcu.h
+++ b/mt76_connac_mcu.h
@@ -1272,6 +1272,7 @@ enum {
 	MCU_EXT_CMD_GROUP_PRE_CAL_INFO = 0xab,
 	MCU_EXT_CMD_DPD_PRE_CAL_INFO = 0xac,
 	MCU_EXT_CMD_PHY_STAT_INFO = 0xad,
+	MCU_EXT_CMD_SET_QOS_MAP = 0xb4,
 	MCU_EXT_CMD_WF_RF_PIN_CTRL = 0xbd,
 };
 
--- a/mt7915/main.c
+++ b/mt7915/main.c
@@ -209,7 +209,7 @@ static int mt7915_add_interface(struct i
 	struct mt7915_phy *phy = mt7915_hw_phy(hw);
 	struct mt76_txq *mtxq;
 	bool ext_phy = phy != &dev->phy;
-	int idx, ret = 0;
+	int idx, i, ret = 0;
 
 	mutex_lock(&dev->mt76.mutex);
 
@@ -257,6 +257,22 @@ static int mt7915_add_interface(struct i
 	mvif->sta.wcid.tx_info |= MT_WCID_TX_INFO_SET;
 	mt76_wcid_init(&mvif->sta.wcid, phy->mt76->band_idx);
 
+	/* init Default QoS map, defined in section 2.3 of RFC8325.
+	 * Three most significant bits of DSCP are used as UP.
+	 */
+	for (i = 0; i < IP_DSCP_NUM; ++i)
+		mvif->qos_map[i] = i >> 3;
+
+	/* Recommended QoS map, defined in section 4 of RFC8325.
+	 * Used in cfg80211_classify8021d since kernel v6.8.
+	 */
+	mvif->qos_map[10] = mvif->qos_map[12] = mvif->qos_map[14] = mvif->qos_map[16] = 0;
+	mvif->qos_map[18] = mvif->qos_map[20] = mvif->qos_map[22] = 3;
+	mvif->qos_map[24] = 4;
+	mvif->qos_map[40] = 5;
+	mvif->qos_map[44] = mvif->qos_map[46] = 6;
+	mvif->qos_map[48] = 7;
+
 	mt7915_mac_wtbl_update(dev, idx,
 			       MT_WTBL_UPDATE_ADM_COUNT_CLEAR);
 
--- a/mt7915/mt7915.h
+++ b/mt7915/mt7915.h
@@ -86,6 +86,8 @@
 
 #define MT7915_RTS_LEN_THRES		0x92b
 
+#define IP_DSCP_NUM			64
+
 struct mt7915_vif;
 struct mt7915_sta;
 struct mt7915_dfs_pulse;
@@ -177,6 +179,8 @@ struct mt7915_vif {
 
 	struct ieee80211_tx_queue_params queue_params[IEEE80211_NUM_ACS];
 	struct cfg80211_bitrate_mask bitrate_mask;
+	/* QoS map support */
+	u8 qos_map[IP_DSCP_NUM];
 };
 
 /* crash-dump */
